### PR TITLE
Refactor/Move qt_plotting to pyvista_qt

### PR DESCRIPTION
**Pull request by akaszynski**
### Move qt_plotting to pyvista_qt

This PR moves `qt_plotting.py` out of `pyvista.plotting` and into [pyvistaqt](https://github.com/pyvista/pyvistaqt).  Reasoning for this:

##### Licensing
GPL scares me a little, and I want to make certain that we've removed all traces of `pyqt` from our module over the concern of licensing.

##### Agile Development
Testing is really bogging down agile development.  It's difficult to get core tests/doc building/ and qt tests to consistently run each time.  It's usually a flaky test in Mac or some similar timeout in Windows due to server load.

With split modules, we can add a ton of features like better icons, `pyqt5` version control, feature tree, etc. all within the setup.py.

##### Better testing control
Changes to `BackgroundPlotter` do not affect any upstream core functionality and as such we should not need to test the core functionality again with each minor change.  The opposite doesn't hold true, (i.e. changes to the `pyvista` core can affect `pyvistaqt`) so we simply pull from master and check if we're valid.

##### Pyside
We can support pyside in the same manner; by providing a submodule that builds upon `BasePlotter`.  Including `pyside` support within `pyvista` will further bloat the module and we need to move the opposite direction.

### Summary
This is a significant but necessary change to make `pyvista` easier to develop in an agile manner.  It will also lead to improvements in `BackgroundPlotter` as changes there will be easier to test and develop.

### CODE REVIEW

**Comment by imsodin** [comment]
In terms of licensing this does not help: Just as currently importing pyqt5 means the whole of pyvista needs to conform with gpl, so does importing pyvista_qt, which in turn imports pyqt5. This problem can only be solved by not importing pyqt5 at all in the codebase.

**Comment by akaszynski** [comment]
> In terms of licensing this does not help: Just as currently importing pyqt5 means the whole of pyvista needs to conform with gpl, so does importing pyvista_qt, which in turn imports pyqt5. This problem can only be solved by not importing pyqt5 at all in the codebase.

It's a reasonable point.  Do you propose we move out the entire import?  As it stands it's a conditional import that only runs if you have `pyvista_qt` installed.

**Comment by imsodin** [comment]
I am not the right person to ask this, I am neither an expert on licensing nor affected by the licensing choice of pyvista. From what I read my (!) conclusion is that importing a python module is considered linking, and linking GPL code makes the whole thing subject to GPL terms. So the question here is: Is conditional importing still importing? I'd guess yes. However I'd expect not even experts could be entirely sure, because these cases usually only become clear after litigation. Legal advice will probably stay on the save side usually. The actual judgement would have to be done by downstreams in violation of GPL (e.g. any closed-source downstreams distributing the code), i.e. whether to use pyvista or not and accept any risk, or pyqt5, i.e. whether or not to litigate. Pyvista itself as far as I understand is not in violation of GPL (unless stating the license to be MIT is already considered a violation - it's probably not correct, but by making the pyvista source open it still complies with GPL as far as I understand). So my personal opinion is: Do not import pyqt5 in any way if you want to license pyvista as MIT. However my opinion is not what counts. I seem to remember that closed-source downstreams have been mentioned at some point. If that's correct whoever knows them should ask them, because their opinion matters.

**Comment by akaszynski** [comment]
Removing the `BackgroundPlotter` all together makes sense to me as it completely removes any doubt that pyqt is or isn't being imported.  Plus, from a development standpoint I don't have to test an additional dependency.

**Comment by banesullivan** [comment]
I vote for taking all traces of PyQt out of `pyvista` (this repo) and into a separate repo with a compatible license. We need to focus on the core feature dev and testing to really move the software forward. Also, having things more modular will streamline dev all around and make it easier for new developers to jump on board with PyVsita core.

I've **majorly ranted** about GPL on many other issues and how we can't even import GPL software without violating the license - so in brief, let's get PyQt out of here and stop worrying about it. Also, having it here is just getting messy.

-----

In practice, we'll need to keep an empty `BackgroundPlotter` and `QtInteractor<whatever_its_called>` class around that throw big ol' warnings. I have a feeling this might be a challenge/(pain in the 🍑) for downstream projects like MNE-Python and the now several commercial software packages using the PyQt interface, so we'll definitely want to have all of them (@GuillaumeFavelier/@larsoner and @math-artist are the first to come to mind) fully in the loop before we even think about merging this.

**Comment by akaszynski** [comment]
> In practice, we'll need to keep an empty `BackgroundPlotter` and `QtInteractor<whatever_its_called>` class around that throw big ol' warnings.

Were you thinking of a depreciation warning?  We can do that for a minor release, but I'd honestly prefer just forcing people off rather than taking the bandaid off slowly.  I've got a closed source project that uses `pyvista.BackgroundPlotter` and something would generally break between versions anyway, so I had to fix the version until I was ready for a true upgrade.  It will be less to maintain as well.

I'd say throwing a depreciation error would be best, but I'll settle with a warning if @GuillaumeFavelier and @larsoner think it's best.

**Comment by akaszynski** [comment]
@banesullivan, I'm going to need you to go into add qtdocs.pyvista.org to [pyvista_qt-docs](https://github.com/pyvista/pyvista_qt-docs). GitHub pages is already enabled, but I need you to setup the domain (or give me access to it).

**Comment by GuillaumeFavelier** [comment]
This is a big change, this plotter is here since I learned about `pyvista`. I think it's enough for a minor release, what do you think @pyvista/developers ?

I thought the plan was to move towards [pyvista-gui](https://github.com/pyvista/pyvista-gui) though. Is it for even more modularity?

Also, I understand it's called `pyvista-qt` with a `-` right? That's what I saw on PyPi. Is there a reason why we should keep the spelling `pyvista_qt` (with a `_`) around? Sorry for the naive question, I just think it should be consistent everywhere.

> I'd say throwing a depreciation error would be best

I'm in favour of the deprecation error. If it's not there anymore, people should not expect it to work. And we can inform the users about the move. Then this can be removed at the next big release cycle. But if we go for a warning, we should redirect to `Plotter` as an alternative.

>  I have a feeling this might be a challenge/(pain in the peach) for downstream projects like MNE-Python and the now several commercial software packages using the PyQt interface

Regarding `mne-python`, the plan is to pin the latest version of `pyvista` with the `BackgroundPlotter` around and work the transition when everything is settled on the new `pyvista-qt` side. I remain optimistic, hopefully everything will go smoothly :+1: 

You might be interested @larsoner, @agramfort :)

**Comment by imsodin** [comment]
I'd definitely consider this a minor bump, it's not just a bugfix. Also this seems like a good time to introduce RCs, given there seems to be quite a few affected downstreams that would benefit from early testing.

The underscore vs hyphen thing is a typical problem: Module names must not use hyphens but underscores (if at all), package names should avoid underscores. I.e. the package would usually be pyvista-qt, the module pyvista_qt. Or just make both pyvistaqt to remove any possible confusion: https://www.python.org/dev/peps/pep-0008/#package-and-module-names

**Comment by larsoner** [comment]
> I remain optimistic, hopefully everything will go smoothly

Indeed I don't foresee any big issues for MNE, we'll just need to implement some `try/except` import patterns at our end to accommodate older and newer styles I think

**Comment by akaszynski** [comment]
> Or just make both pyvistaqt to remove any possible confusion:

I think that sounds like the best solution.  It's not the nicest looking name, but it will lead to the least confusion between `-` and `_`.  I've personally always had issues (especially when downloading from pypi`

> Sorry for the naive question, I just think it should be consistent everywhere.

That's my fault.  Provided no one has an objection with the name `pyvistaqt`, I'll change it shortly.

**Comment by akaszynski** [comment]
> I thought the plan was to move towards [pyvista-gui](https://github.com/pyvista/pyvista-gui) though. Is it for even more modularity?

I think we can slowly add features from `pyvista-gui` into `pyvistaqt`.  We can start with a solid base and then build up from there.

File: docs/getting-started/installation.rst [hunk 1]
```diff
@@ -58,7 +58,7 @@ The following are a list of optional dependencies and their purpose:
 +===================================+=========================================+
 | ``matplotlib``                    | Using Colormaps                         |
 +-----------------------------------+-----------------------------------------+
-| ``PyQt5==5.11.3``                 | Background plotting                     |
+| ``pyvistaqt``                     | Background plotting                     |
```
Comment (banesullivan) [inline docs/getting-started/installation.rst:61]: We should take this out of "dependencies" and just link to it as an offshoot project.

File: pyvista/plotting/__init__.py [hunk 2]
```diff
@@ -11,6 +11,29 @@
...
-from .qt_plotting import BackgroundPlotter, QtInteractor, MainWindow, Counter
 from .helpers import plot, plot_arrows, plot_compare_four, plot_itk
 from .itkplotter import PlotterITK
+
+
+
+class DeprecationError(Exception):
+    """Depreciation Error."""
+
+    pass
```
Comment (banesullivan) [inline pyvista/plotting/__init__.py:22]: Can we add a more informative message by default and call this `QtDeprecationError`

```py
class QtDeprecationError(Exception):
    """Depreciation Error for features that moved to `pyvistaqt`."""
    
    message = """`{}` has moved to pyvistaqt.
    You can install this from PyPI with: `pip install pyvistaqt`
    See https://github.com/pyvista/pyvistaqt
"""
    def __init__(self, feature_name):
        Exception.__init__(self, self.message.format(feature_name))
```

File: pyvista/plotting/plotting.py [hunk 3]
```diff
@@ -3494,8 +3494,8 @@ def orbit():
                 if write_frames:
                     self.write_frame()
 
-
-        if bkg and isinstance(self, pyvista.BackgroundPlotter):
+        is_background_plotter = 'BackgroundPlotter' in str(type(self))
+        if bkg and is_background_plotter:
```
Comment (banesullivan) [inline pyvista/plotting/plotting.py:3498]: This isn't okay. The code here is assuming a feature for a downstream package. The `BackgroundPlotter` downstream should override this method with this background threading feature

Comment (banesullivan) [reply pyvista/plotting/plotting.py:3498]: We should also double-check that we don't do anything similar in other places. A simple grep for `BackgroundPlotter` showed nothing else but something in the back of my mind says there may be another splot in `BasePlotter` (or its inherited classes) where we check if it is a `BackgroundPlotter`

Comment (akaszynski) [reply pyvista/plotting/plotting.py:3498]: Agreed.  I'll patch this.

File: pyvista/plotting/__init__.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/plotting/__init__.py:22]: That's more informative.  I'll patch that.

File: docs/getting-started/installation.rst (hunk 1 above)
Comment (akaszynski) [reply docs/getting-started/installation.rst:61]: Done, and added a reference to `pyvistaqt` in our plotting docs.

**Comment by akaszynski** [comment]
> These changes warrant a MINOR version bump and details should be the top point in the release notes.

Agreed.  I've added it to the milestone.  We've been due for a minor release anyway.

**Pull request merged**